sklearn/decomposition/tests/test_truncated_svd.py

```diff
@@ -220,3 +220,79 @@ def test_singular_values():
     rpca.fit(X_hat_rpca)
     assert_array_almost_equal(apca.singular_values_, [3.142, 2.718, 1.0], 14)
     assert_array_almost_equal(rpca.singular_values_, [3.142, 2.718, 1.0], 14)
+
+
+def test_svd_whiten():
+    tsvd = TruncatedSVD(n_components=X.shape[1] - 1, random_state=42,
+                        whiten=True, algorithm='arpack')
+    Xt = tsvd.fit_transform(X)
+    Xt_2 = tsvd.transform(X)
+    assert_allclose(Xt, Xt_2)
+    Xinv = tsvd.inverse_transform(Xt)
+    # Xdense is a sparse array in dense format (check zeros separately)
+    # The below equality is approximate since n_components < n_features
+    mask = Xdense == 0.0
+    assert_allclose(Xinv[mask], 0, atol=0.08)
+    assert_allclose(Xinv[~mask], Xdense[~mask], rtol=0.04)
+
+
+def test_svd_consistency():
+    X_c = X - X.mean(axis=0)
+    svd = TruncatedSVD(n_components=5, whiten=True, random_state=0)
+    assert_allclose(svd.fit(X_c).explained_variance_,
+                    np.ones(5))
+
+
+def test_truncated_svd_eq_pca():
+    # TruncatedSVD should be equal to PCA on centered data
+
+    X_c = X - X.mean(axis=0)
+
+    pars = dict(n_components=52, random_state=42)
+
+    for whiten in [False, True]:
+        svd = TruncatedSVD(whiten=whiten, **pars)
+        pca = PCA(whiten=whiten, **pars)
+
+        Xt_svd = svd.fit_transform(X_c)
+        Xt_pca = pca.fit_transform(X_c)
+        assert_allclose(Xt_svd, Xt_pca)
+
+        assert_allclose(pca.mean_, 0, atol=1e-9)
+
+        assert_allclose(svd.components_, pca.components_)
+
+        # transform only a subset of the data
+        Xt2_svd = svd.transform(X_c[:10])
+        Xt2_pca = pca.transform(X_c[:10])
+        assert_allclose(Xt2_svd, Xt2_pca)
+
+        X2_svd = svd.inverse_transform(Xt2_svd)
+        X2_pca = pca.inverse_transform(Xt2_svd)
+        assert_allclose(X2_svd, X2_pca)
+
+
+def test_lsi():
+    # Verification of the Latent Semantic Indexing example from
+    # Grossman, Frieder (2004) Information Retrieval, Algorithms and Heuristics
+    # http://www1.se.cuhk.edu.hk/~seem5680/lecture/LSI-Eg.pdf
+    from sklearn.feature_extraction.text import CountVectorizer
+    from sklearn.metrics.pairwise import cosine_similarity
+    documents = ["Shipment of gold damaged in aa fire.",
+                 "Delivery of silver arrived in aa silver truck.",
+                 "Shipment of gold arrived in aa truck."]
+    query = "gold silver truck"
+    n_components = 2
+    # Expected similarity scores between the documents and the query,
+    scores_ref = [-0.0541, 0.9910, 0.4478]
+
+    vect = CountVectorizer()
+    vect.fit(documents)
+    X = vect.transform(documents)
+    q = vect.transform([query])
+    svd = TruncatedSVD(n_components=n_components, whiten=True)
+
+    X_proj = svd.fit_transform(X)
+    q_proj = svd.transform(q)
+    scores = cosine_similarity(X_proj, q_proj)[:, 0]
+    assert_array_almost_equal(scores, scores_ref, decimal=2)
```

Comment:
On my machine this passes with decimal=3 as well. The reference PDF walks through the entire computation using only 4 decimals of precision, so if they are truly being consistent, we cannot expect >3 decimal equality here because their computation is much less numerically precise.

Conclusion: if we pass with decimal=3 on all CIs, I suggest we use that, otherwise 2 is fine.